File: 2167-minimum-time-to-remove-all-cars-containing-illegal-goods/2167-minimum-time-to-remove-all-cars-containing-illegal-goods.cpp
class Solution {
public:
    int minimumTime(string s) {
        int n = s.size();
        vector<int> presum(n);
        presum[0] = (s[0] == '1');
        for(int i = 1; i < n; ++i)
            presum[i] = presum[i-1] + (s[i] == '1');
        vector<int> arr(n);
        for(int i = 0; i < n; ++i)
            arr[i] = 2*presum[i] - i;
        int preMax = 1;
        int ret = n;
        for(int i = 0; i < n; ++i) {
            ret = min(arr[i] - preMax + n, ret);
            preMax = max(preMax, arr[i]);
        }
        return ret;
    }
};

File: 2167-minimum-time-to-remove-all-cars-containing-illegal-goods/2167-minimum-time-to-remove-all-cars-containing-illegal-goods.py
class Solution(object):
    def minimumTime(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s)
        pre = [0]*n
        pre[0] = int(s[0] == '1')
        for i in range(n):
            pre[i] = pre[i-1] + int(s[i] == '1')
        arr = [0]*n
        for i in range(n):
            arr[i] = 2*pre[i] - i
        preMax = 1
        ret = n
        for i in range(n):
            ret = min(ret, arr[i] - preMax + n)
            preMax = max(arr[i], preMax)
        return ret